go/signatures_test.go

```diff
@@ -0,0 +1,310 @@
+	return sigTestSCTWithSignature(t, sigTestCertSCTSignatureRSA, sigTestKeyIDEC)
+}
+
+func sigTestECPublicKey(t *testing.T) crypto.PublicKey {
+	pk, err := PublicKeyFromPEM([]byte(sigTestEC256PublicKeyPEM))
+	if err != nil {
+		t.Fatalf("Failed to parse sigTestEC256PublicKey: %v", err)
+	}
+	return pk
+}
+
+func sigTestRSAPublicKey(t *testing.T) crypto.PublicKey {
+	pk, err := PublicKeyFromPEM([]byte(sigTestRSAPublicKeyPEM))
+	if err != nil {
+		t.Fatalf("Failed to parse sigTestRSAPublicKey: %v", err)
+	}
+	return pk
+}
+
+func sigTestCertLogEntry(t *testing.T) LogEntry {
+	return LogEntry{
+		Index: 0,
+		Leaf: MerkleTreeLeaf{
+			Version:  V1,
+			LeafType: TimestampedEntryLeafType,
+			TimestampedEntry: TimestampedEntry{
+				Timestamp: sigTestSCTTimestamp,
+				EntryType: X509LogEntryType,
+				X509Entry: mustDehex(t, sigTestDERCertString),
+			},
+		},
+	}
+}
+
+func mustCreateSignatureVerifier(t *testing.T, pk crypto.PublicKey) SignatureVerifier {
+	sv, err := NewSignatureVerifier(pk)
+	if err != nil {
+		t.Fatalf("Failed to create SignatureVerifier: %v", err)
+	}
+	return *sv
+}
+
+func TestVerifySCTSignatureEC(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestECPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTEC(t), sigTestCertLogEntry(t)); err != nil {
+		t.Fatalf("Failed to verify signature on SCT: %v", err)
+	}
+
+}
+
+func TestVerifySCTSignatureRSA(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestRSAPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTRSA(t), sigTestCertLogEntry(t)); err != nil {
+		t.Fatalf("Failed to verify signature on SCT: %v", err)
+	}
+
+}
+
+func TestVerifySCTSignatureFailsForMismatchedSignatureAlgorithm(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestECPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTRSA(t), sigTestCertLogEntry(t)); err == nil {
+		t.Fatalf("Sucessfully verified with mismatched signature algorithm")
+	}
+
+}
+
+func TestVerifySCTSignatureFailsForUnknownSignatureAlgorith(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestECPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTWithSignature(t, sigTestCertSCTSignatureUnsupportedSignatureAlgorithm, sigTestKeyIDEC), sigTestCertLogEntry(t)); err == nil {
+		t.Fatal("Successfully verified signature with unsupported signature algorithm")
+	}
+}
+
+func TestVerifySCTSignatureFailsForUnknownHashAlgorith(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestECPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTWithSignature(t, sigTestCertSCTSignatureUnsupportedHashAlgorithm, sigTestKeyIDEC), sigTestCertLogEntry(t)); err == nil {
+		t.Fatal("Successfully verified signature with unsupported hash algorithm")
+	}
+}
+
+func testVerifySCTSignatureFailsForIncorrectInput(t *testing.T, sv SignatureVerifier) {
+	entry := sigTestCertLogEntry(t)
+	entry.Leaf.TimestampedEntry.X509Entry[mrand.Intn(len(entry.Leaf.TimestampedEntry.X509Entry))] ^= byte(mrand.Intn(255) + 1)
```

Comment:
I normally corrupt each byte in a loop in case there's an uncovered byte in there somewhere (likely to be at the ends, of course).